src/components/tl/mlx5/alltoall/alltoall_mkeys.c

```diff
@@ -0,0 +1,448 @@
+                                              void *mkey_entries, int n_entries,
+                                              ucc_base_lib_t *lib)
+{
+    struct ibv_qp_ex *   qp_ex = ibv_qp_to_qp_ex(umr_qp);
+    struct mlx5dv_qp_ex *mqp   = mlx5dv_qp_ex_from_ibv_qp_ex(qp_ex);
+    ucc_status_t         status;
+
+    ibv_wr_start(qp_ex);
+    qp_ex->wr_id = 1; // First (and only) WR
+    mlx5dv_wr_mr_list(mqp, mkey, mem_access_flags, n_entries,
+                      (struct ibv_sge *)mkey_entries);
+
+    if (ibv_wr_complete(qp_ex)) {
+        tl_error(lib, "UMR WQE failed (errno=%d)", errno);
+        return UCC_ERR_NO_MESSAGE;
+    }
+    status = poll_umr_cq(umr_cq, lib);
+    if (status != UCC_OK) {
+        return status;
+    }
+    return UCC_OK;
+}
+
+// Execute the UMR WQE for populating the UMR's MasterMKey
+static ucc_status_t populate_strided_mkey(ucc_tl_mlx5_a2a_t * a2a,
+                                          int                 mem_access_flags,
+                                          struct mlx5dv_mkey *mkey,
+                                          void *mkey_entries, int repeat_count,
+                                          ucc_base_lib_t *lib)
+{
+    ucc_status_t            status;
+    ucc_tl_mlx5_a2a_net_t * net  = &a2a->net;
+    ucc_tl_mlx5_a2a_node_t *node = &a2a->node;
+
+    ucc_tl_mlx5_post_umr(net->umr_qp, mkey, mem_access_flags, repeat_count,
+                         node->sbgp->group_size,
+                         (struct mlx5dv_mr_interleaved *)mkey_entries,
+                         node->umr_entries_mr->lkey, node->umr_entries_buf);
+
+    status = poll_umr_cq(net->umr_cq, lib);
+    if (status != UCC_OK) {
+        tl_error(lib, "failed to populate strided UMR mkey (errno=%d)", errno);
+        return status;
+    }
+    return UCC_OK;
+}
+
+static ucc_status_t create_and_populate_recv_team_mkey(ucc_tl_mlx5_a2a_t *a2a,
+                                                       ucc_base_lib_t *   lib)
+{
+    ucc_status_t            status;
+    ucc_tl_mlx5_a2a_node_t *node = &a2a->node;
+    int                     mnc  = a2a->max_num_of_columns;
+
+    int i, j;
+
+    status = create_master_key(MAX_OUTSTANDING_OPS * mnc, a2a->pd,
+                               &node->team_recv_mkey, lib);
+    if (status != UCC_OK) {
+        return status;
+    }
+    struct ibv_sge *team_mkey_klm_entries = (struct ibv_sge *)calloc(
+        MAX_OUTSTANDING_OPS * mnc, sizeof(struct ibv_sge));
+
+    for (i = 0; i < MAX_OUTSTANDING_OPS; i++) {
+        for (j = 0; j < mnc; j++) {
+            team_mkey_klm_entries[(i * mnc) + j].addr = 0;
+            //length could be minimized for all mkeys beside the first, but no need because address space is big enough
+            team_mkey_klm_entries[(i * mnc) + j].length =
+                node->sbgp->group_size * a2a->max_msg_size * a2a->team_size;
+            //todo check lkey or rkey
+            team_mkey_klm_entries[(i * mnc) + j].lkey =
```

Comment:
calculate and save once per j: index = (i * mnc) + j and then use team_mkey_klm_entries[index].addr/length/lkey